File: examples/structured_mapping/structured_mapping.ml
type t =
  | Lit of float
  | Var of string
  | Plus of t list
  | Exp of t * t
  | Frac of frac

and frac = {num: t; den: t}


let rec eval e env =
  match e with
  | Lit f -> f
  | Var x -> List.assoc x env
  | Plus l -> List.fold_left (+.) 0. (List.map (fun x -> eval x env) l)
  | Exp (x, y) -> eval x env ** eval y env
  | Frac f -> eval f.num env /. eval f.den env



(* Other examples *)

type myrecord = {
  x: int;
  y: myrecord option;
}


type myvariant =
  | A
  | B of string * int
  | C of myvariant

File: examples/structured_mapping/structured_mapping.cs
using System.Collections.Generic;
using LexiFi.Interop;

public static class StructuredMapping {
  public static void Main() {
    LexiFi.Interop.Csml.Init();
    LexiFi.Interop.StructuredMapping.Init();

    /* Support for addin mode. */
    if (!LexiFi.Interop.Csml.MLStubAvailable("StructuredMapping"))
        LexiFi.Interop.Csml.LoadFile("structured_mapping.cma");

    List<Expr> el = new List<Expr>();
    el.Add(new Expr.Var("x"));
    el.Add(new Expr.Var("y"));
    el.Add(Expr.Fraction(new Expr.Lit(1), new Expr.Lit(2)));
    Expr e = new Expr.Plus(el);

    List<Tuple<string,double>> env = new List<Tuple<string,double>>();
    env.Add(new Tuple<string, double>("x", 1));
    env.Add(new Tuple<string, double>("y", 2));
    double x = e.Eval(env);

    System.Console.WriteLine("result = " + x);
  }
}


public class VariantMatching : MyVariant.Match<int> {
  override public int A() { return 1; }
  override public int Bee(string s, int i) { return i; }
  override public int C(MyVariant x) { return this.run(x); }

  public static int Match(MyVariant x) {
    return (new VariantMatching()).run(x);
  }
}
